ACMS - Add helper method to handle the Download Confirmation Page

ERosendo opened the pull request:
This PR adds support to upload documents from the new ACMS website and inserts a banner that lets users know a matching record is available in the RECAP archive.

The PR introduces the following changes:

- Refines the `dispatchHandler` method to effectively manage the **Download confirmation page**. The extension uses the `path` property for page identification.

- Uses the [MutationObserver](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver) interface to monitor DOM changes and ensure all elements are rendered before proceeding. This PR removes the default `onclick` event of the **"Accept Charges"** button and replaces it with a custom event handler that allows us to upload documents to the RECAP archive.

- The Appellate class uses the ACMS API to retrieve the pdf documents so the PR adds a new file named `acms_api.js` to encapsulate utility functions designed to handle API interactions. I used the `exportInstance()` and `importInstance()` methods to enable the communication between the content script and the new service.

- Enhances the `parseReceiptPageTitle` method by refining the regular expressions to extract the data from the title of the download confirmation page. 

- Introduces a new helper function named `storeDownloadDataInSession` that extracts the app settings( API data and Auth token) and the data used by Vue to render the page and stores it in the `sessionStorage` object, allowing the extension access to the same data.

- Add a new parameter to the `uploadDocument` function so we can send the ACMS document GUI when it's available.

ERosendo committed:
feat(appellate): Update the page handler method to identify more pages.

This commit adds the download confirmation page, attachment page and full docket page to the dispatchPageHandler method.

ERosendo committed:
feat(appellate): Add helper method to get data from the download page

johnhawkinson commented on src/appellate/appellate.js:
```diff
@@ -53,7 +60,7 @@ AppellateDelegate.prototype.dispatchPageHandler = function () {
     case 'ShowDocMulti':
       this.handleCombinedPdfPageView();
-      break
+      break;
```
This fixes an older style error (missing ; after break in code added in `66fc121d`). I would not do that in the same commit, but perhaps I am too anal.

This is also not mentioned in the commit message, not even obliquely.

johnhawkinson replied:
6d39debc8dfdbaa7eb99cf6a994d94f12a96921c continues to mix the correction to old code (missing ; after break from 66fc121d43a1708b8412c512d2ea31744f8bace7 from July) with the whitespace corrections that are properly fixups to b4a187017752183bbfafc75eaa125027039485cc (which was never a publishable commit anyhow). Maybe I am the only one who cares.

ERosendo replied:
I'm sorry, I can continue splitting the commit into smaller parts

johnhawkinson commented on src/appellate/appellate.js:
```diff
@@ -103,7 +110,7 @@ AppellateDelegate.prototype.handleAcmsDocket = async function () {
         console.log(` Child nodes: ${t.childNodes.length}`);
       }
     };
-  
```
Separate commits for style/whitespace corrections from previously.
These, in particular, will never be committed to the mainline, so they are different from fixes to prior errors that have seen the light of day.

So it is not just a style preference, but actually history cleanliness and ease of merge conflict resolution that counsels them being separate commits.

johnhawkinson commented on src/appellate/appellate.js:
```diff
@@ -156,7 +163,7 @@ AppellateDelegate.prototype.handleAcmsDocket = async function () {
     }
   };
-  
+
```
Id.

johnhawkinson commented on src/appellate/appellate.js:
```diff
@@ -74,6 +74,51 @@ AppellateDelegate.prototype.dispatchPageHandler = function () {
+    for (const r of mutationList) {
+      for (const n of r.addedNodes) {
+        if (n.localName === 'div') {
+          /* Insert script to store vue data in the storage */
+          APPELLATE.storeDownloadDataInSession();
+
+          /* Check if the submit button is already created on the page */
+          let acceptChargesButtons = document.getElementsByTagName('button');
+          if (!acceptChargesButtons) {
+            return;
+          }
+
+          /* Remove default listener for the submit button */
+          let submitButton = acceptChargesButtons[0];
+          let clonedSubmitButton = submitButton.cloneNode(true);
+          submitButton.replaceWith(clonedSubmitButton);
+
+          clonedSubmitButton.addEventListener('click', function () {
+            /* Gather data to request PDF file */
+            let queryParameters = new URLSearchParams(window.location.search);
+            let includePageNumbers = queryParameters.get('includePageNumbers') ? true : false;
+            let showPDFHeaderInput = document.getElementById('showPdfHeader').checked;
+            let vueData = JSON.parse(sessionStorage.getItem('downloadConfirmationData'));
+
+            const mergePdfFilesRequest = {
+              mergeScope: 'External',
+              pagination: includePageNumbers,
+              header: showPDFHeaderInput,
+              docketEntryDocuments: vueData.docketEntryDocuments,
+            };
+
+            /* TODO: Use the mergePdfFilesRequest to request the PDF doc */
```
Do we use `/*` comments? I thought our code consistently used `//` comments. Obviously not a big deal, and, I guess, not a problem for eslint. I think jslint enforced Crockford's "use line comments, not block comments" rule, although I'm not 100% why. I think it has to do with the problem of commenting out blocks of commented code.

ERosendo replied:
> I thought our code consistently used // comments.

You're right, I've updated the comment style to use `//` instead of `/*` and maintain consistency with the established conventions.

johnhawkinson commented on src/appellate/appellate.js:
```diff
@@ -74,6 +74,51 @@ AppellateDelegate.prototype.dispatchPageHandler = function () {
+    for (const r of mutationList) {
+      for (const n of r.addedNodes) {
+        if (n.localName === 'div') {
+          /* Insert script to store vue data in the storage */
```
think s/vue/Vue for style; in text not in code?

johnhawkinson replied:
Inserting a `<script>` tag into the ACMS web page in order to have it run our code is…a huge _honking_ step. Accepting without research that it is necessary, it needs to be documented  a bit more clearly than just a casual "Insert script to store vue data in the storage" comment and no reference in the commit message.

johnhawkinson commented on src/appellate/appellate.js:
```diff
@@ -74,6 +74,51 @@ AppellateDelegate.prototype.dispatchPageHandler = function () {
+      for (const n of r.addedNodes) {
+        if (n.localName === 'div') {
+          /* Insert script to store vue data in the storage */
+          APPELLATE.storeDownloadDataInSession();
```
I have some reservations about perpetuating the style of putting helper functions in `util.js` that are only called from `appellate.js`. It feels like a good way to give the reader whiplash and make it really hard to figure out what the code is doing. Is there a good reason to continue to do this?

johnhawkinson commented on src/appellate/appellate.js:
```diff
@@ -74,6 +74,51 @@ AppellateDelegate.prototype.dispatchPageHandler = function () {
+  const wrapperMutationObserver = (mutationList, observer) => {
+    for (const r of mutationList) {
+      for (const n of r.addedNodes) {
+        if (n.localName === 'div') {
```
The criteria in `wrapperMutationObserver` all feel very…loose, here and on line 86.. We do this all on the appearance of any `div`, we assume any `button` is the one we care about.

ERosendo replied:
I've introduced stricter criteria in [f12f6af](https://github.com/johnhawkinson/recap-chrome/pull/1/commits/f12f6afec2dd4847b1d5eb5aec89a849aa66ae73). The new criteria ensures the callback function will run the logic only when the mutation that adds content to the Download Confirmation page is included in the array of `MutationRecord` objects.

Let me know what you think.

johnhawkinson commented on src/appellate/appellate.js:
```diff
@@ -74,6 +74,51 @@ AppellateDelegate.prototype.dispatchPageHandler = function () {
+          /* Insert script to store vue data in the storage */
+          APPELLATE.storeDownloadDataInSession();
+
+          /* Check if the submit button is already created on the page */
+          let acceptChargesButtons = document.getElementsByTagName('button');
```
We call this it a "submit button" in the comment but "acceptChargesButton" in the code.
Which is it?

ERosendo replied:
We should call it `acceptChargesButton` because this button is not submitting a form. it's just a button with an `onclick` event.

ERosendo replied:
I've updated the comment in 6d1c36ced8ccdebbefd61c348594eac46dd26d6e

johnhawkinson commented on src/appellate/appellate.js:
```diff
@@ -74,6 +74,51 @@ AppellateDelegate.prototype.dispatchPageHandler = function () {
+            return;
+          }
+
+          /* Remove default listener for the submit button */
```
We talk about removing the default JS method from ACMS's submit button, but we don't say why, and we don't clarify what it did and whether our substitute does all the things that it would have done. I tend to think we would be better off wrapping its event handler in our own and calling it before/after we did our additional stuff. But it's hard to tell from the code what that would be and whether it's right or not. And we should be able to tell.

I would tend to include the removed method in a comment, if only for understandability. But there may be other ways to gain that understandability.

ERosendo replied:
I've included comments in a36ff733d4fc5f15ce49ea6dd27cfb6ee6eefe45 explaining what the default onclick event does and the  issue I've found while attempting to retrieve the document after the event's completion.

johnhawkinson commented on src/appellate/utils.js:
```diff
@@ -406,4 +406,20 @@ let APPELLATE = {
+      let contentWrapper = document.getElementsByClassName('text-center')[0]
+      let vueMainDiv = contentWrapper.parentElement;
+      let vueDataProperties = vueMainDiv.__vue__._data;
+      sessionStorage.setItem("downloadConfirmationData", JSON.stringify(vueDataProperties));
```
If we're going to be messing with session storage in javascript on someone else's page, we should name our storage elements carefully, and probably use `recap` in the name, rather than just `downloadConfirmationData`, if only for our own sanity.

ERosendo replied:
I agree. I will use recap in the name. Thanks

ERosendo replied:
I've tweaked the name of the storage key in 4f4c8b535925936cf24810a3ee35dd0d77e4b4be

johnhawkinson commented on src/appellate/utils.js:
```diff
@@ -406,4 +406,20 @@ let APPELLATE = {
+      let contentWrapper = document.getElementsByClassName('text-center')[0]
+      let vueMainDiv = contentWrapper.parentElement;
+      let vueDataProperties = vueMainDiv.__vue__._data;
+      sessionStorage.setItem("downloadConfirmationData", JSON.stringify(vueDataProperties));
```
Single-quoted strings only, eslint will complain

ERosendo replied:
You're right. I will run Prettier to format this code.

johnhawkinson commented on src/appellate/utils.js:
```diff
@@ -406,4 +406,20 @@ let APPELLATE = {
+  // Adds the vue data attributes to the session storage
+  storeDownloadDataInSession: () => {
+    var code = '(' + function() {
+      let contentWrapper = document.getElementsByClassName('text-center')[0]
```
Missing semicolon

johnhawkinson commented on src/appellate/utils.js:
```diff
@@ -406,4 +406,20 @@ let APPELLATE = {
+
+  // Adds the vue data attributes to the session storage
+  storeDownloadDataInSession: () => {
+    var code = '(' + function() {
+      let contentWrapper = document.getElementsByClassName('text-center')[0]
+      let vueMainDiv = contentWrapper.parentElement;
+      let vueDataProperties = vueMainDiv.__vue__._data;
+      sessionStorage.setItem("downloadConfirmationData", JSON.stringify(vueDataProperties));
+    } + ')();';
+
+    var script = document.createElement('script');
```
`const` and `let` over `var`

johnhawkinson commented on src/appellate/appellate.js:
```diff
@@ -74,6 +74,51 @@ AppellateDelegate.prototype.dispatchPageHandler = function () {
+          clonedSubmitButton.addEventListener('click', function () {
+            /* Gather data to request PDF file */
+            let queryParameters = new URLSearchParams(window.location.search);
+            let includePageNumbers = queryParameters.get('includePageNumbers') ? true : false;
```
I think we'd want to use the `!!foo` idiom rather than `foo?true:false`

ERosendo replied:
You're right!. I'll tweak that line. Thanks

ERosendo replied:
I've added this suggestion in commit d200dc849bae3a4db070c796ec4aeb2426a0ac1c

johnhawkinson commented on src/appellate/utils.js:
```diff
@@ -406,4 +406,20 @@ let APPELLATE = {
+      sessionStorage.setItem("downloadConfirmationData", JSON.stringify(vueDataProperties));
+    } + ')();';
+
+    var script = document.createElement('script');
+    script.textContent = code;
+    (document.head||document.documentElement).appendChild(script);
+
+    script.remove();
```
It's probably no big deal, but is it proper to do this?

At first I thought we were `remove()`ing the `script` element because we were only temporarily creating it in our own document in order to shove it somewhere else, but then on further review I don't think that's right. But then…why are we removing it after we append it? And we're willing to append it in up to two places?
I guess I'm just not understanding what this code is actually doing.

ERosendo replied:
> why are we removing it after we append it?

We're removing the script because It's a small piece of code, We just need it once and the inline script that's inserted in the document is immediately executed and the `<script>` tag can safely be removed. 

> we're willing to append it in up to two places?

We can place script tags inside the head or body section, the `document.head||document.documentElement` returns just one element.

johnhawkinson replied:
I remain confused. I don't mean to create a mountain out of a molehill.

> We're removing the script because It's a small piece of code, We just need it once and the inline script that's inserted in the document is immediately executed and the `<script>` tag can safely be removed.

But this is true of all scripts, is it not?
Any javascript in the DOM of any webpage could remove itself from the DOM after it is executed.
The result would be making it dramatically harder for people to debug and figure out what was going on (or rather, what had happened).

I guess if that's a concern maybe the same logic that animated 3368c54a946ead4c86de1d95cfb91e87d546a3a9 applies and perhaps we should

```
script.dataset.recap = 'Inserted by RECAP Extension.';
```

but maybe that's excessive.

I don't know that it matters much here, but I do not think our threat model (as it were) involves evading any checks for injected Javascript that the ACMS web page might do? So it seems like this removal is unnecessary, potentially confusing, and not standard practice. So why do it?

> > we're willing to append it in up to two places?
> 
> We can place script tags inside the head or body section, the `document.head||document.documentElement` returns just one element.

Yes, but _why_? Is this the standard thing that people do? Is it to overcome some kind of browser compatibility problem? Why not just use `document.head` unconditionally? I don't have a lot of experience with injecting script tags into DOMs, but I haven't been able to find examples that do it this way. What would be necessary for there to be no `document.head`?

johnhawkinson reviewed (commented):
Hooray! Does this code actually handle downloads?

Anyhow, as to the `1d96f42` commit message:
• Commit messages shouldn't start with "This commit [does] …", because if we did that, all commit messages would do that and it would be pointless.

Lots of little nits below.

ERosendo committed:
style(appellate): Add a missing semicolon in the PageHandler method

ERosendo committed:
feat(appellate): Tighten the criteria in the wrapperMutationObserver

Tightens the criteria in wrapperMutationObserver to prevent unnecessary re-renders and improve performance. Previously, the callback would run the logic to remove the event listener from the button and gather the data to upload the PDF whenever any div element was added to the page, even if irrelevant to the extension. This could lead to performance problems, especially when adding banners.

The new criteria ensures that the callback function will run the logic only when the mutation that adds content to the Download Confirmation page is included in the array of MutationRecord objects.

ERosendo committed:
docs(appellate): Update comment about the button on the download page

Use "accept charges" instead of "submit" button to refer to the button on the download page.

ERosendo committed:
refactor(appellate): Tweak the logic to get the accept charges btn

ERosendo committed:
docs(appellate): Adds comments about removing the charges btn default event

The comments provide details regarding the default event handler's responsibilities, outline the issue with the onclick event, and identify the elements included in the newly implemented onclick event.

ERosendo committed:
feat(appellate): Tweaks the helper method to store the ACMS config

Saves the _model attribute to the sessionStorage object.

ERosendo committed:
refactor(appellate): Move onclick event handler logic into a function.

ERosendo committed:
feat(appellate): Update the docId using the data from the sessionStorage

ERosendo committed:
feat(appellate): Inserts the recap banner when a doc is available

ERosendo committed:
style(appellate): Fix styling in the MutationObserver method

ERosendo committed:
feat(appellate): Get ACMS API endpoint and token from the sessionStorage

ERosendo committed:
feat(appellate): Add a helper method to build the PDF request's body

ERosendo committed:
feat(appellate): Adds support to parse ACMS download pages' titles

ERosendo committed:
feat(appellate): Extract case_id and doc GUID from the sessionStorage

ERosendo committed:
feat(appellate): Collect relevant data of the document

ERosendo committed:
feat(appellate): Add utils to retrieve file GUID and compute PDF URL

ERosendo committed:
feat(background): Make the ACMS service callable from content scripts

ERosendo committed:
feat(appellate): Adds an ACMS service instance into the Appellate class

ERosendo committed:
feat(recap): Adds a new parameter to the uploadDocument method

Enhance the uploadDocument function by introducing an optional parameter for the document GUID.

ERosendo committed:
feat(manifest): Adds a pattern to match ACMS website and insert CSS

johnhawkinson reviewed (commented):
I thought we were putting ACMS guids directly into Courtlistener `pacer_case_id`s? Am I confused?

ERosendo commented:
> I thought we were putting ACMS guids directly into Courtlistener pacer_case_ids?

@johnhawkinson You're right. The extension uses the `sessionStorage` object to retrieve the ACMS case ID.

```javascript
// this is a GUID
this.pacer_case_id = downloadData.caseSummary.caseDetails.caseId;

```

ERosendo committed:
feat(appellate): Display a loading message while retrieving the  doc

ERosendo committed:
refactor(appellate): Cast query parameter as a boolean.

Use the double negation operator(!!foo) instead of the ternary conditional operator(foo?true:false) to check the value of the includePageNumbers parameter.
